### PR TITLE
Refine the rebalance scope calculating logic in the WAGED rebalancer.

### PR DESCRIPTION
**Pull request by jiajunwang**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

#518 

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

Refine the rebalance scope calculating logic in the WAGED rebalancer.
1. Ignore the IdealState mapping/listing fields if the resource is in FULL_AUTO mode.
2. On IdealState change, the resource shall be fully rebalanced since some filter conditions might be changed. Such as instance tag.
3. Live instance change (node newly connected) shall trigger full rebalance so partitions will be re-assigned to the new node.
4. A new baseline calculation no longer triggers all the replicas in the cluster to be rebalanced by default. The partial rebalance method will determine the scope based on the real change type.
5. Modify the related test cases.

### Tests

- [x] The following tests are written for this issue:

The changed logic is covered by existing tests.

- [x] The following is the result of the "mvn test" command on the appropriate module:

[ERROR] Failures: 
[ERROR]   TestJobQueueCleanUp.testJobQueueAutoCleanUp » ThreadTimeout Method org.testng....
[ERROR]   TestTaskPerformanceMetrics.testTaskPerformanceMetrics:118 expected:<true> but was:<false>
[ERROR]   TestGetLastScheduledTaskExecInfo.testGetLastScheduledTaskExecInfo:73 expected:<COMPLETED> but was:<RUNNING>
[INFO] 
[ERROR] Tests run: 1043, Failures: 3, Errors: 0, Skipped: 3
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------

Rerun and these 3 tests pass.

### Commits

- [x] My commits all reference appropriate Apache Helix GitHub issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation in the following wiki page:

NA

### Code Quality

- [x] My diff has been formatted using helix-style.xml

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java [hunk 1]
```diff
@@ -97,27 +97,32 @@ public static ClusterModel generateClusterModel(ResourceControllerDataProvider d
...
+    // newly connected nodes = changed liveInstance nodes & currently active instances.
+    Set<String> newlyConnectedNodes = clusterChanges
+        .getOrDefault(HelixConstants.ChangeType.LIVE_INSTANCE, Collections.emptySet());
+    newlyConnectedNodes.retainAll(liveInstances);
     if (clusterChanges.containsKey(HelixConstants.ChangeType.CLUSTER_CONFIG) || clusterChanges
-        .containsKey(HelixConstants.ChangeType.INSTANCE_CONFIG)) {
-      // If the cluster topology has been modified, need to reassign all replicas
+        .containsKey(HelixConstants.ChangeType.INSTANCE_CONFIG) || !newlyConnectedNodes.isEmpty()) {
+      // 1. If the cluster topology has been modified, need to reassign all replicas.
+      // 2. If any node was re-connect, need to reassign all replicas for balance.
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java]: reconnected.
"for balance"? What do you mean exactly? Sorry this is confusing. For global baseline calculation or balance as in evenness?

Follow-up question: this is for recovery from a temporary assignment state (due to delayed rebalance) to the original assignment, correct?

File: helix-core/src/test/java/org/apache/helix/controller/rebalancer/waged/TestWagedRebalancer.java [hunk 2]
```diff
@@ -379,13 +379,13 @@ public void testRebalanceOnChanges() throws IOException, HelixRebalanceException
 
     // 2. rebalance with one ideal state changed only
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/controller/rebalancer/waged/TestWagedRebalancer.java]: Is this "2." still valid? Now we seem to have a ResourceConfig with change.

File: helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeSnapshot.java [hunk 3]
```diff
@@ -72,6 +73,17 @@
...
     _idealStateMap = new HashMap<>(dataProvider.getIdealStates());
+    for (String resourceName : _idealStateMap.keySet()) {
+      IdealState orgIdealState = _idealStateMap.get(resourceName);
+      if (orgIdealState.getRebalanceMode().equals(IdealState.RebalanceMode.FULL_AUTO)) {
+        IdealState trimmedIdealState = new IdealState(orgIdealState.getRecord());
+        // For FullAuto resources, map fields and list fields in the IdealStates is not user's input.
+        // So there is no need to detect any change in these 2 scopes.
+        trimmedIdealState.getRecord().setListFields(Collections.emptyMap());
+        trimmedIdealState.getRecord().setMapFields(Collections.emptyMap());
+        _idealStateMap.put(resourceName, trimmedIdealState);
```
[the remark this answers is not in the document]
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeSnapshot.java]: I think this is a generic logic. Do you have any example that we may need to check map/list fields change when the resource is fullauto? High-levelly, we don't need to check any change that is done by the controller.

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java [hunk 4]
```diff
@@ -69,6 +69,7 @@
   // contains 1. baseline recalculate, 2. partial rebalance that is based on the new baseline.
   private static final Set<HelixConstants.ChangeType> GLOBAL_REBALANCE_REQUIRED_CHANGE_TYPES =
```
[the remark this answers is not in the document]
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java]: Let's worry about namings in the other PR. I don't want to mix unrelated changes.
Let me create a new issue based on this.

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java [hunk 5]
```diff
@@ -97,27 +97,32 @@ public static ClusterModel generateClusterModel(ResourceControllerDataProvider d
...
-   * 4. The resource idealstate has been updated. TODO remove this condition when all resource configurations are migrated to resource config.
-   * 5. If the current best possible assignment does not contain the partition's valid assignment.
+   * 2. The resource config has been updated.
+   * 3. If the current best possible assignment does not contain the partition's valid assignment.
    *
    * @param replicaMap             A map contains all the replicas grouped by resource name.
    * @param clusterChanges         A map contains all the important metadata updates that happened after the previous rebalance.
-   * @param activeInstances        All the instances that are alive and enabled.
+   * @param activeInstances        All the instances that are alive and enabled according to the delay rebalance configuration.
+   * @param liveInstances          All the instances that are alive.
```
[the remark this answers is not in the document]
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java]: 1. the znode is called liveInstance
2. we don't use aliveInstance in our code elsewhere. I believe all of the vars are named as "liveNode" or "liveInstance".

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java [hunk 6]
```diff
@@ -254,8 +255,6 @@ public void close() {
     if (clusterChanges.keySet().stream()
         .anyMatch(GLOBAL_REBALANCE_REQUIRED_CHANGE_TYPES::contains)) {
       refreshBaseline(clusterData, clusterChanges, resourceMap, currentStateOutput);
-      // Inject a cluster config change for large scale partial rebalance once the baseline changed.
-      clusterChanges.putIfAbsent(HelixConstants.ChangeType.CLUSTER_CONFIG, Collections.emptySet());
```
[the remark this answers is not in the document]
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java]: It is hacky, and I want to remove it.
Let me add it to the description. It is also related to the scope thing.

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java (hunk 1 above)
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java]: I forgot to change the comments after renamed the parameter. Should be "newly connected".
It's not confusing, it is for both. I guess you know too many details about the internal implementation so you are confused. The logic here does not care about global or partial. This is just for generically balancing the replicas for evenness.

For the follow-up, maybe for recovery, maybe for new nodes, this logic does not really care. New nodes, then try to move everything. That's it.

File: helix-core/src/test/java/org/apache/helix/controller/rebalancer/waged/TestWagedRebalancer.java (hunk 2 above)
Comment (jiajunwang) [reply helix-core/src/test/java/org/apache/helix/controller/rebalancer/waged/TestWagedRebalancer.java]: Will update.

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java [hunk 7]
```diff
@@ -97,27 +97,32 @@ public static ClusterModel generateClusterModel(ResourceControllerDataProvider d
...
    */
   private static Set<AssignableReplica> findToBeAssignedReplicas(
       Map<String, Set<AssignableReplica>> replicaMap,
       Map<HelixConstants.ChangeType, Set<String>> clusterChanges, Set<String> activeInstances,
-      Map<String, ResourceAssignment> bestPossibleAssignment,
+      Set<String> liveInstances, Map<String, ResourceAssignment> bestPossibleAssignment,
       Map<String, Set<AssignableReplica>> allocatedReplicas) {
     Set<AssignableReplica> toBeAssignedReplicas = new HashSet<>();
+
+    // newly connected nodes = changed liveInstance nodes & currently active instances.
```
[the remark this answers is not in the document]
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java]: changed liveInstance nodes  => changed liveInstance znodes

What we need is finding the new live nodes.
Another way I tried is using the detector to find added nodes. But then how to find a node that is reconnected with a different session? In that case, we also need to treat it as a newly added nodes in the rebalancer. So the current method is the easiest way. Although there might be some false positive, it's still better than missing one new node adding event.

File: helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeSnapshot.java [hunk 8]
```diff
@@ -72,6 +73,17 @@
     _changedTypes = new HashSet<>(dataProvider.getRefreshedChangeTypes());
     _instanceConfigMap = new HashMap<>(dataProvider.getInstanceConfigMap());
     _idealStateMap = new HashMap<>(dataProvider.getIdealStates());
+    for (String resourceName : _idealStateMap.keySet()) {
+      IdealState orgIdealState = _idealStateMap.get(resourceName);
+      if (orgIdealState.getRebalanceMode().equals(IdealState.RebalanceMode.FULL_AUTO)) {
+        IdealState trimmedIdealState = new IdealState(orgIdealState.getRecord());
+        // For FullAuto resources, map fields and list fields in the IdealStates is not user's input.
+        // So there is no need to detect any change in these 2 scopes.
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeSnapshot.java]: Nit: these 2 scopes -> this would technically be "these two fields"?

File: helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeSnapshot.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (narendly) [reply helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeSnapshot.java]: Sure, what you are saying is correct. But ChangeDetector is a public component, and we can't make the assumption that the ListField and mapField are populated only by the controller. Users may actually use IdealState's list and map fields, and they are very much allowed to use ChangeDetector to build some custom logic around it (although that's certainly not how we intended). What do you think?

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (narendly) [reply helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java]: I saw the issue for this. Acknowledged.

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (narendly) [reply helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java]: Sounds good. I thought it was a hack. Please include it in the description.

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (narendly) [reply helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java]: What I was referring to was the description:

"All the instances that are alive" -> "All the instances that are live". "Live" would be proper since instances are not living organisms :)

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (narendly) [reply helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java]: That's perfectly fine, but could you please add that information as an inline comment? 

"changed" is vague. Could you please add the part about LiveInstances having a different sessionId? That's much easier to understand :)

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java (hunk 1 above)
Comment (narendly) [reply helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java]: In that case, could we rephrase it to "for evenness of distribution" or something like that? Or better, say that "we want to reassign so that some replicas would be placed on newly connected instances for better evenness".

I know this is nitpicky, but it really helps if I know what's going on clearly :)

Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java]: I refine the statement. Thanks for pointing this out.

File: helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeSnapshot.java (hunk 8 above)
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeSnapshot.java]: Good point.

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java]: Modified.

File: helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeSnapshot.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeSnapshot.java]: This is a good example. I think we can add more checks to ensure the trim is done safely.
But back to the original question, adding options or make the logic into the rebalancer could be very confusing design. I would still prefer to leave the logic inside as a default behavior.
Please check my latest update.

**Comment by jiajunwang** [comment]
> I know this is nitpicky, but it really helps if I know what's going on clearly :)

@narendly No worry, nitpicky is totally fine.
However, I do wish that we can focus more on code logic. For example, in the "trimmedIdealState" part, it was problematic that I only trimmed for FULL_AUTO mode only, but not SEMI_AUTO. When you said this should be generic, it reminded me. But it would be even better if you, as the reviewer, can find these problems. There could be more, you know : )
Again, I think your comments are good ones. And thanks for reviewing in detail.

**Comment by jiajunwang** [comment]
Discussed with @narendly offline, it would be a good idea to add an option to the change detector for genericity. The user can choose if the detector should detect for Helix source change or not.

File: helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeDetector.java [hunk 9]
```diff
@@ -50,8 +51,13 @@
   private Map<HelixConstants.ChangeType, Collection<String>> _addedItems = new HashMap<>();
   private Map<HelixConstants.ChangeType, Collection<String>> _removedItems = new HashMap<>();
 
-  public ResourceChangeDetector() {
+  public ResourceChangeDetector(boolean ignoreHelixSourceChange) {
     _newSnapshot = new ResourceChangeSnapshot();
+    _ignoreHelixSourceChange = ignoreHelixSourceChange;
+  }
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeDetector.java]: @jiajunwang This is not what we agreed on during our offline discussion. I think the idea was to extend ResourceChangeDetector (for example, ControllerResourceChangeDetector or WagedRebalancerChangeDetector), that simply overrides a method and includes the filtering/trimming logic.

Adding a field like `_ignoreHelixSourceChange` *was* an option that we discussed, but we both agreed that this would make the code not very clean. Could we please try the subclass method like we agreed to earlier?

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java [hunk 10]
```diff
@@ -449,7 +448,7 @@ private void refreshBaseline(ResourceControllerDataProvider clusterData,
 
   private ResourceChangeDetector getChangeDetector() {
     if (CHANGE_DETECTOR_THREAD_LOCAL.get() == null) {
-      CHANGE_DETECTOR_THREAD_LOCAL.set(new ResourceChangeDetector());
+      CHANGE_DETECTOR_THREAD_LOCAL.set(new ResourceChangeDetector(true));
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java]: ResourceChangeDetector's constructor should not change. Please see the comment above.

File: helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeDetector.java (hunk 9 above)
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeDetector.java]: I tried. You know I always try if there is a plan B before sending PR.
So here let's call what I did plan A. Then plan B would be creating a class called something like ControllerResourceChangeDetector, as we discussed. However, when there are 2 classes exist in the project, the users might be confused, they will need to read the description of these classes then decide which one to use. Since the option in plan A explains itself good enough, I feel this way is much easier for those who may use this class in the future.
Moreover, it is not straightforward to describe what is a "ControllerResourceChangeDetector" and explain why we need an additional class.  I'm not a fan of increasing documentation while we can let the code explain for itself. 
Finally, logically, I feel Helix controller should not have a special requirement to the change detector. Or our design might be not good enough in some way.

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java]: Sure, will modify.

File: helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeSnapshot.java [hunk 11]
```diff
@@ -113,4 +123,26 @@
...
+
+  // Trim the IdealState to exclude any controller modified information.
+  private IdealState trimIdealState(IdealState originalIdealState) {
+    // Clone the IdealState to avoid modifying the objects in the Cluster Data Cache, which might
+    // be used by the other stages in the pipeline.
+    IdealState trimmedIdealState = new IdealState(originalIdealState.getRecord());
+    switch (originalIdealState.getRebalanceMode()) {
+      case FULL_AUTO:
+        // For FULL_AUTO resources, both map fields and list fields are not considered as data input
+        // for the controller.
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeSnapshot.java]: Also add:
The controller will write to mapFields and listFields meaning they will be overwritten by the controller...

File: helix-core/src/test/java/org/apache/helix/controller/changedetector/TestResourceChangeDetector.java [hunk 12]
```diff
@@ -338,6 +342,47 @@ public void testNoChange() {
     }
   }
 
+  /**
+   * Modify IdealState mapping fields for a FULL_AUTO resource and see if detector detects.
+   */
+  @Test//(dependsOnMethods = "testNoChange")
+  public void testIgnoreHelixSourceChange() {
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/controller/changedetector/TestResourceChangeDetector.java]: testIgnoreControllerGeneratedFieldChange?

File: helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeSnapshot.java [hunk 13]
```diff
@@ -67,27 +68,36 @@
    * Constructor using controller cache (ResourceControllerDataProvider).
    *
    * @param dataProvider
+   * @param ignoreHelixSourceChange if true, the snapshot won't record any changes that is modifying
```
Comment (i3wangyi) [inline helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeSnapshot.java]: nit: grammar error

File: helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeDetector.java [hunk 14]
```diff
@@ -50,8 +51,13 @@
...
+
+  public ResourceChangeDetector() {
```
[the remark this answers is not in the document]
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeDetector.java]: If you don't have this constructor, user has to specify ResourceChangeDetector(false).
Note the default constructor will be invalid in Java if there is any other constructor. So we need to implement the default one.

**Comment by jiajunwang** [comment]
This PR is ready to be merged, approved by @narendly
